Fix providesModuleNodeModules behavior to be more deterministic.

skevy opened the pull request:
--- Relates to https://github.com/facebook/react-native/pull/5985 ---

Currently, the `providesModuleNodeModules` option allows for specifying an array of package names, that the packager will look for within node_modules, no matter how deep they're nested, and treat them as packages that use the `@providesModule` pragma.

In reality, this is not actually the behavior we want.

npm, because of how it handles dependencies, can do all kinds of arbitrary nesting of packages when `npm install` is run. This causes a problem for how we deal with `providesModuleNodeModules`. Specifically...take `fbjs`. Currently, React Native relies on using the Haste version of `fbjs` (will be resolved in #5084). Thus if npm installs multiple copies of fbjs...which is a very common thing for it to do (as can be seen by this list of issues: https://github.com/facebook/react-native/issues?utf8=%E2%9C%93&q=naming+collision+detected), we get into a state where the packager fails and says that we have a naming collision.

Really, the behavior we want is for the packager to treat only a *single* copy of a given package, that we specify in the `Resolver` in the `providesModuleNodeModules` option, as the package that it uses when trying to resolve Haste modules.

This PR provides that behavior, by changing `providesModuleNodeModules` from a list of strings to a list of objects that have a `name` key, specifying the package name, as well as a `parent` key. If `parent` is null, it will look for the package at the top level (directly under `node_modules`). If `parent` is specified, it will use the package that is nested under that parent as the Haste module.

To anyone who reads this PR and is familiar with the differences between npm2 and npm3 -- this solution works under both, given the fact that we are now shipping the NPM Shrinkwrap file with React Native when it's installed through `npm`. In both the npm2 and npm3 case, node_modules specified by RN's package.json are nested underneath `react-native` in node_modules, thus allowing us to specify, for example, that we want to use React Native's copy of `fbjs` (not any other copy that may be installed) as the module used by the packager to resolve any `requires` that reference a module in `fbjs`.

skevy commented:
cc @cpojer @martinbigio

cpojer commented:
NICE

cpojer commented on src/DependencyGraph/DependencyGraphHelpers.js:
```diff
@@ -8,26 +8,83 @@
+   * or examples).
+   */
+  _resolveHastePackages(packages) {
+    const _packagePathForPackage = ({ name, parent }, rootDir) => {
```
can you drop the leading underscore? :)

cpojer commented on src/DependencyGraph/DependencyGraphHelpers.js:
```diff
@@ -8,26 +8,83 @@
+          parent = [parent];
+        }
+        parent.push(name);
+        return path.join(rootDir, 'node_modules', parent.join(`${path.sep}node_modules${path.sep}`)) + path.sep;
```
to me it makes sense not to use template literals here.

cpojer commented on src/DependencyGraph/DependencyGraphHelpers.js:
```diff
@@ -8,26 +8,83 @@
-    for (let i = 0; i < dirs.length; i++) {
-      if (parts.indexOf(dirs[i]) > -1) {
-        return false;
+    const isHaste = this._providesModuleNodeModuleDirs.map(dir => {
```
so the reason why I rewrote this method recently is because it is *super hot* code. We call it 50k times on our react-native codebase during haste map initialization. Can you optimize this code? It seems to me that you can use a for loop and `return true` the first time it matches?

cpojer commented on src/DependencyGraph/DependencyGraphHelpers.js:
```diff
@@ -8,26 +8,83 @@
-        return false;
+    const isHaste = this._providesModuleNodeModuleDirs.map(dir => {
+      if (file.startsWith(dir) &&
+          !file.startsWith(path.join(dir, 'node_modules'))
```
I'd like to avoid calling `path.join` here for perf reasons. Can we make sure that `dir` never ends in a path-separator in `_resolveHastePackages` and use `+ path.sep +` instead?

cpojer commented on src/DependencyGraph/DependencyGraphHelpers.js:
```diff
@@ -8,24 +8,89 @@
+        packagePath = path.join(rootDir, 'node_modules', name);
+      }
+
+      if (packagePath[packagePath.length - 1] === path.sep) {
```
I think `.endsWith` here would be fine too :P

davidaurelio replied:
If I read this correctly, this can only ever be true if name ends with `path.sep`. Does that happen or is this defensive programming? Maybe it’s better sanitizing at the top of the function?

skevy replied:
This can happen in the case of a providesModuleNodeModule entry that looks like:

```{ name: 'some-module', parent: ['parentPackage1', 'parentPackage2'] }```, as the path will end with `/node_modules/`

cpojer commented:
Alright, looks great to me. I'd be happy to merge. I think it makes sense to add one or two tests for `DependencyGraphHelpers` so that we can make sure this behavior will continue to work when we touch this code again?

davidaurelio commented on src/DependencyGraph/DependencyGraphHelpers.js:
```diff
@@ -8,24 +8,89 @@
+   */
   isNodeModulesDir(file) {
-    const index = file.lastIndexOf('/node_modules/');
+    const index = file.indexOf('/node_modules/');
```
``file.indexOf(`${path.sep}node_modules${path.sep}`);``? that’s so ugly that you might want to put a constant at the top of the file :-)

skevy replied:
:+1:

davidaurelio commented on src/__mocks__/graceful-fs.js:
```diff
@@ -112,6 +112,47 @@ fs.stat.mockImpl(function(filepath, callback) {
+    node = getToNode(filepath);
+  } catch (e) {
+    throw e;
+  }
```
Pointless try/catch, unless I’m missing something?

skevy replied:
Yah, you're right, in this context it's not necessary because this is a sync method. I had copied this from the `fs.stat` mock, which WOULD need a try/catch because it needs to errBack.

davidaurelio commented on src/__tests__/DependencyGraph-test.js:
```diff
@@ -90,14 +90,14 @@ describe('DependencyGraph', function() {
-        'react-native',
+        { name: 'haste-fbjs', parent: null },
+        { name: 'react-haste', parent: null },
+        { name: 'react-native', parent: null },
```
Why do we need support for `parent` at all?

skevy replied:
We need support for this use case:

```
- node_modules
  - react-native
    - node_modules
      - fbjs
  - react
    - node_modules
      - fbjs
```

If we're not checking for the `providesModuleNodeModule`'s parent module (as we weren't before), this will break. What this PR does is allow us to specify that we absolutely want to use the fbjs that's a dep of react-native...not the fbjs that's a peer of react-native (in the case of npm3 and Relay) or fbjs that is a dep of React.

As you know https://github.com/facebook/react-native/pull/5084 makes the fbjs use-case moot...but that's not merged yet, and perhaps there could be other cases in the future where this is necessary.

Anyway, as I explained in my PR description, it seems that the current behavior is the _wrong_ behavior, as it's not at all deterministic, which is exactly the problem (or one of the problems haha) that Haste aims to solve.

davidaurelio commented:
Thank you for this PR!
Will this also work for cases like

```
project-dir
└ node_modules
  ├ react
  │ └ node_modules
  │   └ fbjs@0.6.0
  └ react-native
    └ node_modules
      └ fbjs@0.7.1
```
?

Add some meaningful tests, please, and make sure you actually see them fail once without your code changes.

davidaurelio commented:
As @cpojer mentioned, this code has some hefty micro optimizations, especially around the `path` module.

consider turning these:

```js
parent.join(path.sep + 'node_modules' + path.sep)
// …
packagePath = path.join(rootDir, 'node_modules', name);
// …
const index = file.indexOf('/node_modules/');
```

into 

```js
const NODE_MODULES = path.sep + 'node_modules' + path.sep;
// …
parent.join(NODE_MODULES)
// …
packagePath = rootDir + NODE_MODULES + name;
// …
const index = file.indexOf(NODE_MODULES);
```

skevy commented:
@davidaurelio yep that case is exactly what this PR is meant to address -- as I say in my PR description, the issue with how this was being handled before was that the providesModuleNodeModules option wasn't deterministic enough to guarantee that we were looking for the correct version of a package that was listed (such as fbjs)...it would just treat all of versions that it found as Haste packages, resulting in duplicate errors. This PR resolves that.

Per both you and @cpojer's comments, I will add a few tests (and will make sure they're valid tests haha).

Also, I've already noticed the perf issue and am in the process of profiling the new impl of this method against the old one, and am determined to at least make it match the old perf (if not better it slightly).

skevy commented:
I spoke to @cpojer offline about what I'm working on re this PR as well...if it weren't for silly time zones it probably would have been done before you even got a chance to do your review of this :-p

Thanks for your comments :-)

skevy commented:
Just updated stuff @davidaurelio @cpojer.

I made the fixes requested in both of your comments and also added tests (both for DependencyGraphHelpers directly as well as another test in DependencyGraph-tests...which I confirmed was broken with the current code in master).

In addition, I was able to make a pretty significant perf improvement.

Instead of looping through and doing string comparisons in `isNodeModulesDir`, I instead precompile a regex that allows for testing all `providesModuleNodeModules` at once.

![perf](http://s.sk3vy.com/fCIZ/Screen%20Shot%202016-02-18%20at%204.51.20%20PM.png)

I'm able to get a > 90% perf improvement on that method using the regex technique. I also confirmed that this perf improvement exists on both Node 4 and 5.

cpojer commented on src/DependencyGraph/DependencyGraphHelpers.js:
```diff
@@ -8,26 +8,115 @@
-      if (parts.indexOf(dirs[i]) > -1) {
-        return false;
-      }
+    if (this._hasteRegex.test(file)) {
```
`return !this._hasteRegex.test(file);` ?

skevy replied:
Yes. Definitely. My bad.

cpojer commented on src/DependencyGraph/DependencyGraphHelpers.js:
```diff
@@ -8,26 +8,115 @@
+   * during `isNodeModulesDir`, which is run tens of thousands of times in a typical
+   * project. A regex is much faster in this use-case.
+   */
+  _buildHasteRegex(dirs) {
```
I'd probably take this off the class and make it a module-local function. What do you think?

skevy replied:
I'm cool with that.

cpojer commented on src/DependencyGraph/DependencyGraphHelpers.js:
```diff
@@ -8,26 +8,115 @@
+      const dirRegex = '((?!' +
+        escapeRegExp(dir + NODE_MODULES_DIR) + ')' +
+        escapeRegExp(dir + path.sep) + ')';
+      regexes.push(dirRegex);
```
not sure you are getting much mileage out of this reduce, forEach + an outer `regexes` variable seems like one line less and added readability? :P

skevy replied:
Nobody likes `.reduce`! I <3 `.reduce`.

But yes, I can do a `forEach` here, np.

cpojer commented:
Nice, thanks for the work and for making the perf better :) I'll defer to @davidaurelio to take another look and merge it.

skevy commented:
Ok, got rid of the `.reduce`'s @cpojer for readability's sake.

davidaurelio commented:
@skevy thank you for taking the time to wrap this up. Lgtm. The leading underscore convention for module-private functions looks a little awkward to me, but I can live with it.

I’m just. figuring out the best order to merge all open PRs. Patch-level stuff first. Given that this changes the format of `providesModuleNodeModules`, it will be v3.0.0.

Your patch has conflicts.

cpojer commented:
@skevy are you rebasing this so @davidaurelio can merge it?

skevy commented:
@davidaurelio @cpojer sorry for the delay on this. was traveling. Fixed conflicts and got rid of that underscore in the module private function. Should be good to go.

davidaurelio commented:
Published as v2.6.0

cpojer commented:
This PR completely breaks Jest 0.9 unfortunately. We need to revert and re-publish or fix this ASAP (and the next version of Jest should lock node-haste to ~ instead of ^). Pretty sure the same thing will come up in react-native too.

See https://github.com/facebook/jest/issues/771 – the git repo included ( https://github.com/AdamKyle/Toy-Box ) shows well how it breaks when you remove everything inside of `modulePathIgnorePatterns`.
